File: FrogInMaze/FrogInMaze.py
from collections import defaultdict

def isValid(x, y):
    return 0 <= x < n and 0 <= y < m

n, m, k = map(int, input().split())
maze = [list(input()) for _ in range(n)]
tunnel = {}
for _ in range(k):
    x1, y1, x2, y2 = map(int, input().split())
    tunnel[(x1-1, y1-1)] = (x2-1, y2-1)
    tunnel[(x2-1, y2-1)] = (x1-1, y1-1)

# MDP Variables
states = defaultdict()
states[0] = (-1, 1)
state = 1

for i, x in enumerate(maze):
    for j, y in enumerate(x):
        if y in '%AO':
            states[(i, j)] = state
            state += 1
            if y == 'A':
                startState = state - 1

probabilities = [0.0]
t = [[(0, 1.0)]]

for i, x in enumerate(maze):
    for j, y in enumerate(x):
        if y in '#*':
            continue
        elif y == '%':
            probabilities.append(1.0)
            t.append([(states[(i, j)], 1.0)])
        else:
            probabilities.append(0.0)
            if (i, j) in tunnel:
                ni, nj = tunnel[(i, j)]
            else:
                ni, nj = i, j
            success, fail = [], 0
            for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                if isValid(ni + dx, nj + dy):
                    if maze[ni + dx][nj + dy] in 'AO%':
                        success.append((ni + dx, nj + dy))
                    elif maze[ni + dx][nj + dy] == '*': fail += 1
            if not success:
                t.append([(0, 1.0)])
            else:
                temp = []
                for s in success:
                    temp.append((states[s], 1/(len(success) + fail)))
                if fail > 0:
                    temp.append((0, fail/(len(success) + fail)))
                t.append(temp)

minDifference = 10**-10
minDifferenceReached = False

while not minDifferenceReached:
    prob = probabilities.copy()
    for s in range(state):
        probState = 0
        for a, b in t[s]:
            probState += probabilities[a] * b
        probabilities[s] = probState
    minDifferenceReached = True
    for s in range(state):
        if abs(probabilities[s] - prob[s]) > minDifference:
            minDifferenceReached = False
            break

print(probabilities[startState])   

'''
3 6 1
###*OO
O#OA%O
###*OO
2 3 2 1
'''

'''
3 7 0
##OOO##
%OOAOO*
##OOO##
'''

'''
3 5 0
OO#OO
AO#O%
OO#OO
'''

File: FrogInMaze/FrogInMaze.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>

using namespace std;

bool isValid(int x, int y, int n, int m) {
    return 0 <= x && x < n && 0 <= y && y < m;
}

int main() {
    int n, m, k;
    cin >> n >> m >> k;
    vector<string> maze(n);
    for(int i = 0; i < n; i++)
        cin >> maze[i];
    map<pair<int, int>, pair<int, int>> tunnel;
    for(int i = 0; i < k; i++) {
        int x1, y1, x2, y2;
        cin >> x1 >> y1 >> x2 >> y2;
        tunnel[{x1-1, y1-1}] = {x2-1, y2-1};
        tunnel[{x2-1, y2-1}] = {x1-1, y1-1};
    }

    map<pair<int, int>, int> states;
    states[{-1, 1}] = 0;
    int state = 1, startState;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            if(maze[i][j] == 'A' || maze[i][j] == 'O' || maze[i][j] == '%') {
                states[{i, j}] = state++;
                if(maze[i][j] == 'A')
                    startState = state - 1;
            }
        }
    }

    vector<double> probabilities(state, 0.0);
    vector<vector<pair<int, double>>> t(state);
    t[0].push_back({0, 1.0});

    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            if(maze[i][j] == '#' || maze[i][j] == '*')
                continue;
            if(maze[i][j] == '%') {
                probabilities[states[{i, j}]] = 1.0;
                t[states[{i, j}]].push_back({states[{i, j}], 1.0});
                continue;
            }
            int ni = i, nj = j;
            if(tunnel.count({i, j}))
                tie(ni, nj) = tunnel[{i, j}];
            vector<pair<int, int>> success;
            int fail = 0;
            for(auto [dx, dy] : vector<pair<int, int>>{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}) {
                if(isValid(ni + dx, nj + dy, n, m)) {
                    if(maze[ni + dx][nj + dy] == 'A' || maze[ni + dx][nj + dy] == 'O' || maze[ni + dx][nj + dy] == '%')
                        success.push_back({ni + dx, nj + dy});
                    else if(maze[ni + dx][nj + dy] == '*')
                        fail++;
                }
            }
            if(success.empty())
                t[states[{i, j}]].push_back({0, 1.0});
            else {
                for(auto [x, y] : success)
                    t[states[{i, j}]].push_back({states[{x, y}], 1.0 / (success.size() + fail)});
                if(fail > 0)
                    t[states[{i, j}]].push_back({0, double(fail) / (success.size() + fail)});
            }
        }
    }

    double minDifference = 1e-10;
    bool minDifferenceReached = false;

    while(!minDifferenceReached) {
        vector<double> prob = probabilities;
        for(int s = 0; s < state; s++) {
            double probState = 0;
            for(auto [a, b] : t[s])
                probState += probabilities[a] * b;
            probabilities[s] = probState;
        }
        minDifferenceReached = true;
        for(int s = 0; s < state; s++) {
            if(abs(probabilities[s] - prob[s]) > minDifference) {
                minDifferenceReached = false;
                break;
            }
        }
    }

    cout << probabilities[startState] << endl;

    return 0;
}
